### ISSUE TITLE
Wrong response code for creating entities without id or with id = 0

### ISSUE THREAD

**Issue by Wilt** [zfcampus/zf-rest#82]
Wrong response code is returned when creating entities with id = 0 or entities without an id (composite key) because of an if clause in [line 384 in the RestController](https://github.com/zfcampus/zf-rest/blob/master/src/RestController.php#L384):

```
    if ($entity->id) {
        $self = $entity->getLinks()->get('self');
        $self = $plugin->fromLink($self);
        $response = $this->getResponse();
        $response->setStatusCode(201);
        $response->getHeaders()->addHeaderLine('Location', $self);
    }
```

What is the purpose of this if clause? Can it perhaps be replaced by the following:
```
if ($entity instanceof Entity){
   ...
}
```

**Comment by weierophinney** [comment]
I think the check for an identifier is because it's generally required in order to get the `self` relational link. However, I don't see anything in the code that should cause this to break; I'm going to try it now.

**Comment by weierophinney** [comment]
And, actually, testing with `instanceof` is unnecessary, as `createHalEntity()` will only ever return an entity instance.

**Comment by weierophinney** [comment]
@Wilt Okay, there are two related issues:

- #43
- #51 

It looks like the assumption I made in the past is that an error will occur if there is no identifier present in the entity when generating the URI; however, what actually happens is that when the identifier is missing, because it's optional, it's simply dropped. I'm not sure if this is a good or a bad thing; I can certainly see it being unexpected. However, the only case this should typically happen is if you have a composite ID from multiple fields in the entity, which you are likely accounting for via a listener when generating the link.

Thoughts? You can see my attempt here:

- https://github.com/weierophinney/zf-rest/commit/48dfb23267b1999542128d9a555f983df65469f6

**Comment by weierophinney** [comment]
@Wilt I think I've figured out another way to accomplish this, while working on #79.

**Comment by weierophinney** [comment]
@Wilt The way to fix this was to check for a `self` relational header, which you can see here:

- https://github.com/zfcampus/zf-rest/pull/84/files#diff-9f89f7db73347387a05cdd06cc52abd8R394

The whole reason for that conditional was to *not* create the `Location` header in the response if we don't have a `self` relational header. The assumption at this point is that if we do, the plugin will be able to render it properly, so I now test only for the existence of that header.

**Comment by Wilt** [comment]
I'm out of office for another two weeks but i promise I'll look into it as soon as I'm back...
Greetings,
Wilt

On Jul 7, 2015 11:46 PM, weierophinney <notifications@github.com> wrote:
@Wilt Okay, there are two related issues:

- #43
- #51

It looks like the assumption I made in the past is that an error will occur if there is no identifier present in the entity when generating the URI; however, what actually happens is that when the identifier is missing, because it's optional, it's simply dropped. I'm not sure if this is a good or a bad thing; I can certainly see it being unexpected. However, the only case this should typically happen is if you have a composite ID from multiple fields in the entity, which you are likely accounting for via a listener when generating the link.

Thoughts? You can see my attempt here:

- https://github.com/weierophinney/zf-rest/commit/48dfb23267b1999542128d9a555f983df65469f6


---

**Comment by weierophinney** [comment]
@Wilt I think I have this resolved now, actually, with the fix for #79.

**Comment by Wilt** [comment]
Looks great. Seems to me it does all that is necessary and it definitely solves my issue.
